### PR TITLE
chore: disallow async locks (RwLock)

### PR DESCRIPTION
**Pull request by rumenov**

### CODE REVIEW

**Review by mraszyk — APPROVED** [review]

**Review by mbjorkqvist — APPROVED** [review]

**Comment by pakhomov-dfinity** [comment]
Is there any context for this change?

**Comment by rumenov** [comment]
> Is there any context for this change?

This is just a style guide update.

 If you read https://tokio.rs/tokio/tutorial/shared-state and https://draft.ryhl.io/blog/shared-mutable-state/ it pretty much argues that we should avoid the tokio::sync::{Mutex|RwLock}.

File: rs/boundary_node/ic_boundary/src/core.rs
Comment (fspreiss) [inline rs/boundary_node/ic_boundary/src/core.rs:1]: [Same comment as before](https://github.com/dfinity/ic/pull/2132#discussion_r1808655207):
> Instead of allowing all disallowed types in the entire file, couldwe apply this allow-rule only to the line(s) where it actually applies? This way we would catch disallowed types in files where other disallowed types are already present. If there are multiple/too many occurrences, maybe an alias type like the one used in line 16 (use tokio::sync::Mutex as AsyncMutex;) could help.

If we leave it to the code owners to clean this up, ideally a ticket is created for them, because without a ticket this will just be forgotten once the PR is merged.

**Review by fspreiss — APPROVED** [review]

File: rs/boundary_node/ic_boundary/src/core.rs
Comment (rumenov) [reply rs/boundary_node/ic_boundary/src/core.rs:1]: https://dfinity.atlassian.net/browse/BOUN-1284
https://dfinity.atlassian.net/browse/CON-1430
https://dfinity.atlassian.net/browse/FI-1541

**Review by schneiderstefan — APPROVED** [review]

**Review by r-birkner — APPROVED** [review]

File: clippy.toml
Comment (daniel-wong-dfinity-org) [inline clippy.toml:15]: You might want to make the reason more prominent. E.g.

> This is slow (compared to what? by how much?). For (better) alternatives, see https://tokio.rs/tokio/tutorial/shared-state. This ban can be overridden by adding `#![allow(clippy::disallowed_types)] // This is an appropriate use of Mutex, because...` to the top of the file where the banned type is used.

I think the rest is not very helpful. E.g. "This usually is not necessary" is not actionable. Also, "You should only use X if you need Y." probably does not help, because presumably, the reason X is being used is that the author believes that they need Y.

File: rs/registry/local_registry/src/lib.rs
Comment (daniel-wong-dfinity-org) [inline rs/registry/local_registry/src/lib.rs:1]: Maybe, all of these legacy ignores should say that they are legacy. Otherwise, people should normally include a comment that explains why the ban does not cover their use case. There shouldn't be bare `allows`.

**Review by daniel-wong-dfinity-org — APPROVED** [review]

File: clippy.toml
Comment (rumenov) [reply clippy.toml:15]: "the reason X is being used is that the author believes that they need Y"

Usually the reason is because the author copied the code. People use the first thing they believe it works without thinking about use cases or alternatives. So it is better not to give much options.

File: rs/registry/local_registry/src/lib.rs
Comment (rumenov) [reply rs/registry/local_registry/src/lib.rs:1]: I am not sure if it matters it is legacy or not. Maybe the default is if there is no comment explaining why the allow is there then the allow should be fixed and used.

**Review by dist1ll — APPROVED** [review]
Thanks for creating the tracking tickets.

**Pull request merged**
